astacus/coordinator/cluster.py

```diff
@@ -198,16 +199,20 @@ async def wait_successful_results(
                     continue
                 progress_text = f"{result.progress!r}" if result is not None else "not started"
                 logger.info("%s node #%d/%d: %s", node_op_from_url(url), i, len(urls), progress_text)
-                r = await utils.httpx_request(
+                async with httpx_request_stream(
                     url, caller="Nodes.wait_successful_results", timeout=self.poll_config.result_timeout
-                )
-                if r is None:
-                    failures[i] += 1
-                    if failures[i] >= self.poll_config.maximum_failures:
-                        raise WaitResultError("too many failures")
-                    continue
-                # We got something -> decode the result
-                result = result_class.parse_obj(r)
+                ) as r:
+                    if r is None:
+                        failures[i] += 1
+                        if failures[i] >= self.poll_config.maximum_failures:
+                            raise WaitResultError("too many failures")
+                        continue
+                    # We got something -> decode the result
+                    assert isinstance(r, httpx.Response)
+                    payload = bytearray()
+                    async for chunk in r.aiter_bytes():
+                        payload.extend(chunk)
+                result = msgspec.json.decode(payload, type=result_class)
```

Comment:
I'm not seeing any performance advantage from using stream here.  We are just collecting into a `bytearray` anyway - might as well let the library do this.

Reply:
Using the stream API reduces the memory usage: without that, the http library does extra copies when merging chunks, and it keeps a separate copy of the entire content attached to the response object.

Reply:
> the http library does extra copies when merging chunks

When using the non stream API it is just doing `self._content = b"".join([part async for part in self.stream])` which is basically just what you've implemented, except maybe more efficient in terms of copying (if `len(part)` is large enough).

https://github.com/encode/httpx/blob/master/httpx/_models.py#L422

There might be a very small memory benefit to your code over theirs since it doesn't have to have an extra array of references to `bytes`.

> and it keeps a separate copy of the entire content attached to the response object

Not a copy just a reference.

Reply:
I got a little confused, because you're right that `_content` is just a reference, but I do see extra memory usage in the profiler without that change.

(This change is actually in a separate commit, because first I tried without, then I saw the `b"".join[...]` still allocated during peak memory usage, and that disappeared once I switched to the stream approach).

I assumed it was because the reference to the response was somehow kept alive for longer than expected (that wouldn't be completely unexpected, especially when you have things like sentry that hook themselves everywhere to collect info about requests). But I was wrong, that's not what happens at all.

After checking the size and number of chunks we have for large response and after digging through cpython, the extend is really more efficient, but what really happens is:

In the `bytearray.extend` case:
 - httpx allocates and fetches a 64k chunk of response in a `bytes`, yields it
 - `bytearray_extend` is passed that chunk, extends its internal buffer with `realloc` (with a multiplicative factor of 1.125, so progressively appending 1000 chunks of 64k does not result in 1000 reallocs, it only does 60 reallocs, and the system `realloc` for large sizes is itself smart about avoiding copies anyway).
 - the 64k chunk is now useless, it has zero refs and can be freed (that memory will be reused for the next block).
 - repeat that for each chunk until the response is complete.

In the `b"".join[...]` case:
 - to allocate the output buffer, `stringlib_bytes_join` needs to know the size of all chunks that it needs to join together.
 - so it has to consume the entire iterable of all chunks.
 - this allocates all N chunks of 64k bytes.
 - then it can allocate a large buffer of N*64k bytes.
 - then it copies everything from all the small chunks to the large buffer.
 - finally it can free all chunks.

So, effectively, the join will use ~twice as much memory, because at some point it will retain both the joined response and all the small chunks at the same time, while extend will only retain one chunk at a time.